### PR TITLE
NODE-2024 Base koa-router transaction names off last matched path

### PR DESCRIPTION
**Pull request by psvet**
## CHANGELOG

* Refactored instrumentation to base transaction names off of [koa-router](https://npmjs.com/package/koa-router)'s matched path when used.

## NOTES

Matching off of last matched path of route declarations (Layers) that can serve requests (`opts.end`) which is mostly consistent with `_matchedRoute`. `_matchedRoute` doesn't currently filter out Layers that do not serve requests but we may submit a PR to resolve that and move to using that eventually.

### CODE REVIEW

File: lib/instrumentation.js [hunk 1]
```diff
@@ -27,16 +27,18 @@ module.exports = function initialize(shim, Koa) {
...
-      var spec = {
+    const stack = router.stack
+    for (let i = 0; i < stack.length; ++i) {
+      const layer = stack[i]
+      const spec = {
         route: layer.path,
         type: shim.MIDDLEWARE,
         next: shim.LAST,
         promise: true,
+        appendPath: false,
```
Comment (michaelgoin) [inline lib/instrumentation.js:41]: I'm surprised doing this across the board for all routers isn't causing issues, such as with `koa-route` tests. Our coverage for those tests seems reasonable though. 🤷‍♀️

**Review by michaelgoin — DISMISSED** [review]
Seems legit. Only surprise is our doing `appendPath: false` across the board not breaking anything with the other routing tests. The default naming must not have been relying on that behavior. Our test coverage there looked good to me on quick scan but may be worth a second scan.

File: lib/instrumentation.js (hunk 1 above)
Comment (psvet) [reply lib/instrumentation.js:41]: I don't think `koa-route` actually uses a `router` prop, so it probably wouldn't touch this. That instrumentation (not to mention the lib itself) is pretty simple by comparison.

**Review by michaelgoin — APPROVED** [review]
👍

**Pull request merged**